balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -384,30 +405,69 @@ func (b *pickfirstBalancer) requestConnectionLocked() {
 		switch scd.state {
 		case connectivity.Idle:
 			scd.subConn.Connect()
+			b.scheduleNextConnectionLocked()
+			return
 		case connectivity.TransientFailure:
-			// Try the next address.
+			// The SubConn is being re-used and failed during a previous pass
+			// over the addressList. It has not completed backoff yet.
+			// Mark it as having failed and try the next address.
+			scd.connectionFailed = true
 			lastErr = scd.lastErr
 			continue
 		case connectivity.Ready:
 			// Should never happen.
 			b.logger.Errorf("Requesting a connection even though we have a READY SubConn")
+			return
 		case connectivity.Shutdown:
 			// Should never happen.
 			b.logger.Errorf("SubConn with state SHUTDOWN present in SubConns map")
+			return
 		case connectivity.Connecting:
 			// Wait for the SubConn to report success or failure.
+			b.scheduleNextConnectionLocked()
+			return
 		}
-		return
 	}
+
 	// All the remaining addresses in the list are in TRANSIENT_FAILURE, end the
-	// first pass.
-	b.endFirstPassLocked(lastErr)
+	// first pass if possible.
+	b.endFirstPassIfPossibleLocked(lastErr)
+}
+
+func (b *pickfirstBalancer) scheduleNextConnectionLocked() {
+	if !envconfig.PickFirstHappyEyeballsEnabled {
+		return
+	}
+	b.cancelScheduled()
+	ctx, cancel := context.WithCancel(context.Background())
+	closeFn := timerAfterFunc(connectionDelayInterval, func() {
+		b.mu.Lock()
+		defer b.mu.Unlock()
+		// If the scheduled task is cancelled while acquiring the mutex, return.
+		if ctx.Err() != nil {
+			return
+		}
+		if b.logger.V(2) {
+			b.logger.Infof("Happy Eyeballs timer expired.")
```

Comment:
Would it be possible to include some other data here, either the pointer to the subchannel or the address of the subchannel, so that the log statement is more meaningful.